src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,323 @@
+
+        final List<SeatNonBid> seatNonBids = auctionContext.getBidRejectionTrackers().entrySet().stream()
+                .map(entry -> toSeatNonBid(entry.getKey(), entry.getValue()))
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .toList();
+
+        final Map<String, NonBid> seatsWithNonBids = Optional.of(seatNonBids)
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatNonBid::getSeat,
+                                seatNonBid -> seatNonBid.getNonBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final CommonMessage commonMessage = createCommonMessage(auctionContext, auctionElapsed);
+
+        return addBidResponseToMessage(
+                commonMessage,
+                imps,
+                seatsWithBids,
+                seatsWithNonBids
+        );
+    }
+
+    @Override
+    public <T> Future<Void> processEvent(T event) {
+        AuctionContext greenbidsAuctionContext = null;
+        BidResponse greenbidsBidResponse = null;
+
+        if (event instanceof AmpEvent ampEvent) {
+            greenbidsAuctionContext = ampEvent.getAuctionContext();
+            greenbidsBidResponse = ampEvent.getBidResponse();
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            greenbidsAuctionContext = auctionEvent.getAuctionContext();
+            greenbidsBidResponse = auctionEvent.getBidResponse();
+        }
+        assert greenbidsBidResponse != null && greenbidsAuctionContext != null;
+
+        this.greenbidsId = UUID.randomUUID().toString();
+        this.billingId = UUID.randomUUID().toString();
+        this.isSampled = isSampled(greenbidsConfig.getGreenbidsSampling(), greenbidsId);
+
+        final CommonMessage commonMessage = createBidMessage(greenbidsAuctionContext, greenbidsBidResponse);
+
+        String commonMessageJson = null;
+        try {
+            commonMessageJson = JsonUtil.toJson(commonMessage);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        HttpUtil.sendJson(commonMessageJson, ANALYTICS_SERVER);
+
+        return Future.succeededFuture();
+    }
+
+    public boolean isSampled(
+            double samplingRate,
+            String greenbidsId
+    ) {
+        if (samplingRate < 0 || samplingRate > 1) {
+            throw new IllegalArgumentException("Warning: Sampling rate must be between 0 and 1");
+        }
+
+        final double exploratorySamplingRate = samplingRate * this.exploratorySamplingSplit;
+        final double throttledSamplingRate = samplingRate * (1.0 - this.exploratorySamplingSplit);
+
+        long hashInt = Math.abs(greenbidsId.hashCode());
+        hashInt = hashInt % 0x10000;
```

Comment:
why not plain integer? why is it calculated in two steps?

Reply:
Concerning the hexadecimal similar as in previous comment. 

Concerning two steps yes I agree, I've modified modulo in a one step